discovery.md

```diff
@@ -0,0 +1,378 @@
+
+There are several discovery use cases to consider from the viewpoint of event
+consumers.
+
+1. What sources are available, and what event types do they produce?
+2. What event types are available, and from which providers.
+
+The second case becomes relevant if multiple providers support the same event
+types. Use case 1 is likely the dominant use case. Given the example of a public
+cloud provider where all producers produce events, there may be dozens of
+sources (producer systems) and hundreds of event types. The discovery funnel of
+producer first, then event types helps users navigate without having to see large
+lists of event types. Both of these cases require that filters can be used
+in the discovery API to narrow down the selection of available events.
+
+The CloudEvent source attribute is a potential source of high fanout. For
+example consider a blob storage system where each directory constitutes a
+distinct `source` attribute value. For this reason, source is not the starting
+point for discovery, but an attribute that can be used in creating subscriptions.
+
+There are use cases where the Discovery API is served as a collection of other
+producers. These take two forms:
+
+1. Middleware
+2. Aggregator (or *Events Broker*)
+
+In the Middleware case, discovery is a combined set of producers for where the
+discovery is passed through largely unmodified and subscriber URIs are generally
+absolute (back to the producer).
+
+In the aggregator use case, the aggregator is expected to modify the output of
+discovery and subscription URIs will be relative to the aggregator itself.
+
+## Notations and Terminology
+
+### Notational Conventions
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+### Terminology
+
+This specification defines the following terms:
+
+### Producer
+
+The "producer" is a specific instance, process or device that creates the data
+structure describing the CloudEvent.
+
+### Intermediary
+
+An "intermediary" receives a message containing an event for the purpose of
+forwarding it to the next receiver, which might be another intermediary or a
+consumer. A typical task for an intermediary is to route the event to receivers
+based on the information in the event context.
+
+### Middleware
+
+Specific to discovery and subscription capabilities. Events middleware systems
+provide discovery services that may be directly passing through discovery from
+producers, or may aggregate or otherwise alter the discovery. This includes
+events brokers that may also be relaying discovery AND subscription information
+further upstream to event producers. Multiple middleware systems may exist
+between the producer and the consumer.
+
+### Consumer
+
+The entity receiving the event - the destination for the events related to a
```

Comment:
Could we just use the CE definition here so that we don't get all tangled-up in the push vs pull aspects?